src/contracts/orderbook/NiftyswapOrderbook.sol

```diff
@@ -0,0 +1,422 @@
+            // Is a listing, cancelled, completed or never existed
+            revert InvalidOfferId(offerId);
+        }
+
+        _acceptOrder(offerId, offer, quantity, additionalFees, additionalFeeRecievers);
+
+        emit OfferAccepted(offerId, msg.sender, offer.tokenContract, quantity);
+    }
+
+    function _acceptOrder(
+        bytes32 orderId,
+        Order memory order,
+        uint256 quantity,
+        uint256[] memory additionalFees,
+        address[] memory additionalFeeRecievers
+    ) internal {
+        if (quantity == 0 || quantity > order.quantity) {
+            revert InvalidQuantity();
+        }
+        if (_isExpired(order)) {
+            revert InvalidExpiry();
+        }
+        if (additionalFees.length != additionalFeeRecievers.length) {
+            revert InvalidAdditionalFees();
+        }
+
+        // Update order state
+        if (order.quantity == quantity) {
+            // Refund some gas
+            delete orders[orderId];
+        } else {
+            orders[orderId].quantity -= quantity;
+        }
+
+        // Calculate payables
+        uint256 remainingCost = order.pricePerToken * quantity;
+        (address royaltyRecipient, uint256 royaltyAmount) =
+            getRoyaltyInfo(order.tokenContract, order.tokenId, remainingCost);
+
+        address currencyReceiver = order.isListing ? order.creator : msg.sender;
+        address tokenReceiver = order.isListing ? msg.sender : order.creator;
+
+        if (royaltyAmount > 0) {
+            // Transfer royalties
+            TransferHelper.safeTransferFrom(order.currency, tokenReceiver, royaltyRecipient, royaltyAmount);
+            remainingCost -= royaltyAmount;
+        }
+
+        // Transfer additional fees
+        uint256 feesPaid = 0;
+        for (uint256 i; i < additionalFees.length; i++) {
+            uint256 fee = additionalFees[i];
+            if (additionalFeeRecievers[i] == address(0) || fee == 0) {
+                revert InvalidAdditionalFees();
+            }
+            TransferHelper.safeTransferFrom(order.currency, tokenReceiver, additionalFeeRecievers[i], fee);
+            if (!order.isListing) {
+                // Fees are paid by the taker. This reduces the cost for offers.
+                remainingCost -= fee;
+            }
+        }
+
+        // Transfer currency
+        TransferHelper.safeTransferFrom(order.currency, tokenReceiver, currencyReceiver, remainingCost);
+
+        // Transfer token
+        address tokenContract = order.tokenContract;
+        if (_tokenIsERC1155(tokenContract)) {
+            IERC1155(tokenContract).safeTransferFrom(currencyReceiver, tokenReceiver, order.tokenId, quantity, "");
+        } else {
+            IERC721(tokenContract).transferFrom(currencyReceiver, tokenReceiver, order.tokenId);
+        }
+    }
+
+    /**
+     * Cancels a listing.
+     * @param listingId The ID of the listing.
+     */
+    function cancelListing(bytes32 listingId) external {
```

Comment:
Why not a single `cancel` method?

Reply:
For constency with `createListing/Offer`, which are different to limit the amount of branching due to different validations.
I'm 50/50 about combining these and using `Order` throughout. What do you think?

Reply:
Updated. See other comment. Resolving this convo